src/main/java/com/icthh/xm/ms/entity/service/ContentService.java

```diff
@@ -1,29 +1,128 @@
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.codec.digest.DigestUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.Pair;
 import org.springframework.stereotype.Service;
 import org.springframework.transaction.annotation.Transactional;
 
+import java.io.IOException;
 import java.util.List;
 
+@Slf4j
 @Service
 @Transactional
 @RequiredArgsConstructor
 public class ContentService {
 
-    private final PermittedRepository permittedRepository;
+    private static final String FILE_NAME_SEPARATOR = "::";
 
-    private final PermittedSearchRepository permittedSearchRepository;
+    private final PermittedRepository permittedRepository;
+    private final ContentRepository contentRepository;
+    private final S3StorageRepository s3StorageRepository;
 
+    @Transactional(readOnly = true)
     @FindWithPermission("CONTENT.GET_LIST")
     @PrivilegeDescription("Privilege to get all the contents")
     public List<Content> findAll(String privilegeKey) {
         return permittedRepository.findAll(Content.class, privilegeKey);
     }
 
+    public Attachment save(AttachmentSpec spec, Attachment attachment, Content content) {
+        // XmEntityServiceImpl.addFileAttachment(XmEntity entity, MultipartFile file) already save file
+        if (attachment.getContentUrl() != null && content == null) {
+            return attachment;
+        }
+
+        if (content == null) {
+            throw new IllegalArgumentException("Attachment content is null!");
+        }
+
+        if (spec.getStoreType() == AttachmentStoreType.S3) {
+            String folderName = attachment.getXmEntity().getTypeKey();
+            UploadResultDto uploadResult = s3StorageRepository.store(content, folderName, attachment.getName());
+            attachment.setValueContentSize((long) content.getValue().length);
+            attachment.setContentChecksum(uploadResult.getETag());
+            attachment.setContentUrl(s3FileName(uploadResult));
+        } else {
+            Content savedContent = contentRepository.save(content);
+            attachment.setContent(savedContent);
+            attachment.setValueContentSize((long) content.getValue().length);
+            attachment.setContentChecksum(DigestUtils.sha256Hex(content.getValue()));
+        }
+
+        return attachment;
+    }
+
+    private String s3FileName(UploadResultDto uploadResult) {
+        return uploadResult.getBucketName() + FILE_NAME_SEPARATOR + uploadResult.getKey();
+    }
+
+    public void delete(Attachment attachment) {
+        if (isS3Compatible(attachment.getContentUrl())) {
+            Pair<String, String> s3BucketNameKey = getS3BucketNameKey(attachment.getContentUrl());
+            s3StorageRepository.delete(s3BucketNameKey.getKey(), s3BucketNameKey.getValue());
+        }
+    }
+
+    @SneakyThrows
+    public Attachment enrichContent(AttachmentSpec spec, Attachment attachment) {
```

Comment:
attachment has typeKey, and spec should be fetched, not passed as parameter to this level

fetch spec by typeKey it's free operation